### ISSUE TITLE
Version 2.00 default oscilloscope app not loading

### ISSUE THREAD

**Issue by werefkin** [RedPitaya/RedPitaya#255]
Hello,
the default oscilloscope app - scopegenpro - not loading in 90% cases, it is independent on the browser. The restarts do not help as well.
What might be an issue and solution?

**Comment by b-swat** [comment]
There is such a problem. It will be fixed in the new version.
The problem itself is that, after loading the page, all the parameters from the board are not requested.

**Comment by PauliusMorku** [comment]
I have the same issue. Seems like there are no commits for past 4 moths, is there anyone still working on this new version?

**Comment by b-swat** [comment]
Adding new functionality, we constantly postpone the new release.
I made a build from the development branch, it fixed bugs with loading applications.

https://downloads.redpitaya.com/downloads/Unify/nightly_builds/

**Comment by spicydonkey** [comment]
@b-swat in the official beta release, is the incomplete load of the device tree overlay a bug with the `jupyter/redpitaya` package?  Specifically, I expect devices like `osc0` and `osc1` to appear under `/dev/uio/` when I load the overlay in python:
`from redpitaya.overlay.mercury import mercury as overlay
fpga = overlay()`

With the old OS, I get:
```
root@rp-f09e25:~# ls -la /dev/uio
total 0
drwxr-xr-x  2 root root  280 Jul  6 04:29 .
drwxr-xr-x 11 root root 3240 Jul  6 04:29 ..
lrwxrwxrwx  1 root root    7 Jul  6 04:21 api -> ../uio0
lrwxrwxrwx  1 root root    7 Jul  6 04:29 clb -> ../uio4
lrwxrwxrwx  1 root root    8 Jul  6 04:29 ctrg -> ../uio11
lrwxrwxrwx  1 root root    7 Jul  6 04:29 gen0 -> ../uio5
lrwxrwxrwx  1 root root    7 Jul  6 04:29 gen1 -> ../uio6
lrwxrwxrwx  1 root root    7 Jul  6 04:29 hwid -> ../uio1
lrwxrwxrwx  1 root root    8 Jul  6 04:29 la -> ../uio10
lrwxrwxrwx  1 root root    7 Jul  6 04:29 lg -> ../uio9
lrwxrwxrwx  1 root root    7 Jul  6 04:29 mgmt -> ../uio2
lrwxrwxrwx  1 root root    7 Jul  6 04:29 osc0 -> ../uio7
lrwxrwxrwx  1 root root    7 Jul  6 04:29 osc1 -> ../uio8
lrwxrwxrwx  1 root root    7 Jul  6 04:29 pdm -> ../uio3
```

But in the beta release, I get a reduced list of devices:
```
root@rp-f09dc4:~# ls -la /dev/uio
total 0
drwxr-xr-x  2 root root  140 Jul  6 04:47 .
drwxr-xr-x 10 root root 3140 Jul  6 04:47 ..
lrwxrwxrwx  1 root root    7 Feb 17 01:43 api -> ../uio0
lrwxrwxrwx  1 root root    7 Jul  6 04:47 clb -> ../uio4
lrwxrwxrwx  1 root root    7 Jul  6 04:47 hwid -> ../uio1
lrwxrwxrwx  1 root root    7 Jul  6 04:47 mgmt -> ../uio2
lrwxrwxrwx  1 root root    7 Jul  6 04:47 pdm -> ../uio3
```

Would you be able to advise whether there is a quick fix to get the `osc` working again?  Does your overnight build fix this issue?

**Comment by b-swat** [comment]
@spicydonkey 

I checked this problem on a new 'Night build', it seems that it was already solved there.

```
Welcome to Ubuntu 22.04.2 LTS (GNU/Linux 5.15.0-xilinx armv7l)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

This system has been minimized by removing packages and content that are
not required on a system that users do not log into.

To restore this content, you can run the 'unminimize' command.
##############################################################################
# Red Pitaya GNU/Linux Ecosystem
# Version: 2.00-27dfa19da
# Build: 128
# Branch: 
# Commit: 27dfa19da1cac87ae64dc6755bb4f729a6ffdfce
# U-Boot: "redpitaya-v2022.1"
# Linux Kernel: "branch-redpitaya-v2022.3"
# Pro Applications: ""
##############################################################################
Last login: Fri Jun 23 13:58:01 2023
root@rp-f01c3d:~# ls -la /dev/uio
total 0
drwxr-xr-x  2 root root   60 Jun 23 13:57 .
drwxr-xr-x 10 root root 3060 Jun 23 13:58 ..
lrwxrwxrwx  1 root root    7 Jun 23 13:57 api -> ../uio0
root@rp-f01c3d:~# overlay.sh mercury
Time taken to load BIN is 222.000000 Milli Seconds
BIN FILE loaded through FPGA manager successfully
root@rp-f01c3d:~# ls -la /dev/uio
total 0
drwxr-xr-x  2 root root  280 Jul  6 08:43 .
drwxr-xr-x 10 root root 3280 Jul  6 08:43 ..
lrwxrwxrwx  1 root root    7 Jun 23 13:57 api -> ../uio0
lrwxrwxrwx  1 root root    7 Jul  6 08:43 clb -> ../uio4
lrwxrwxrwx  1 root root    8 Jul  6 08:43 ctrg -> ../uio11
lrwxrwxrwx  1 root root    7 Jul  6 08:43 gen0 -> ../uio5
lrwxrwxrwx  1 root root    7 Jul  6 08:43 gen1 -> ../uio6
lrwxrwxrwx  1 root root    7 Jul  6 08:43 hwid -> ../uio1
lrwxrwxrwx  1 root root    8 Jul  6 08:43 la -> ../uio10
lrwxrwxrwx  1 root root    7 Jul  6 08:43 lg -> ../uio9
lrwxrwxrwx  1 root root    7 Jul  6 08:43 mgmt -> ../uio2
lrwxrwxrwx  1 root root    7 Jul  6 08:43 osc0 -> ../uio7
lrwxrwxrwx  1 root root    7 Jul  6 08:43 osc1 -> ../uio8
lrwxrwxrwx  1 root root    7 Jul  6 08:43 pdm -> ../uio3

```

**Comment by spicydonkey** [comment]
Thanks, that is a fantastic news!  I will try out the latest nightly build, and report the result.

**Comment by spicydonkey** [comment]
Hello @b-swat , 

I was able to build the nightly build following the documented instructions, and can confirm that the mercury overlay works for me.  I was quite happy at first, but saddened by the following result, that you may know already.  Can you explain what's going on?  But maybe this is derailing this Issue...

I found a degradation in performance against the "stable OS 1.04", with respect to the `osc.data` call that loads the waveform in buffer to python's numpy array.  The call used to take 4 ms, but now it takes 80 ms.  Below is the code and results from each setup:

Here is the minimal viable example python script `time_osc_data.py`:
```
#!/usr/bin/env python3

import time

from redpitaya.overlay.mercury import mercury as overlay
fpga = overlay()
# time.sleep(2)  # this does not affect the duration of `osc.data`

osc0 = fpga.osc(0, 1.0)

# data rate decimation
osc0.decimation = 4
# trigger timing [sample periods]
N = osc0.buffer_size
osc0.trigger_pre  = 0
osc0.trigger_post = N
osc0.trig_src = 0  # disable hardware trigger sources

for i in range(10):
    # synchronization and trigger sources are the default,
    # which is the module itself
    osc0.reset()
    osc0.start()
    osc0.trigger()
    # wait for data
    while osc0.status_run():
        pass
    print ('triggered')

    time_start = time.perf_counter()
    data = osc0.data(N)  # read buffer into numpy array
    time_end = time.perf_counter()
    print('duration: {:.3f} ms'.format((time_end - time_start) * 1000))

```

Here is the result for OS 2.0:
```
Welcome to Ubuntu 22.04.2 LTS (GNU/Linux 5.15.0-xilinx armv7l)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage

This system has been minimized by removing packages and content that are
not required on a system that users do not log into.

To restore this content, you can run the 'unminimize' command.
##############################################################################
# Red Pitaya GNU/Linux Ecosystem
# Version: 2.00-27dfa19da
# Build: 128
# Branch:
# Commit: 27dfa19da1cac87ae64dc6755bb4f729a6ffdfce
# U-Boot: "redpitaya-v2022.1"
# Linux Kernel: "branch-redpitaya-v2022.3"
# Pro Applications: ""
##############################################################################
Last login: Fri Jul  7 01:21:52 2023
root@rp-f09dc4:~# python3 time_osc_data.py
Time taken to load BIN is 273.000000 Milli Seconds
BIN FILE loaded through FPGA manager successfully
Load overlay [OK].
triggered
duration: 77.541 ms
triggered
duration: 80.101 ms
triggered
duration: 80.103 ms
triggered
duration: 79.345 ms
triggered
duration: 79.704 ms
triggered
duration: 80.004 ms
triggered
duration: 79.604 ms
triggered
duration: 79.403 ms
triggered
duration: 79.925 ms
triggered
duration: 79.438 ms
Overlay __del__ was activated.
But since there are garbage collection ordering issues, this code is commented out.
Exception ignored in: <function osc.__del__ at 0xb0081b68>
Traceback (most recent call last):
  File "/home/jupyter/RedPitaya/redpitaya/drv/osc.py", line 62, in __del__
  File "/home/jupyter/RedPitaya/redpitaya/drv/uio.py", line 90, in __del__
BufferError: cannot close exported pointers exist
```

Here is the result for OS 1.04:
```
Welcome to Ubuntu 16.04.7 LTS (GNU/Linux 4.9.0-xilinx armv7l)

 * Documentation:  https://help.ubuntu.com
 * Management:     https://landscape.canonical.com
 * Support:        https://ubuntu.com/advantage
##############################################################################
# Red Pitaya GNU/Linux Ecosystem
# Version: 1.04
# Build: 18
# Branch:
# Commit: 27498667c96692b6f15a7878462480c66c6c0abf
# U-Boot: "redpitaya-v2016.4"
# Linux Kernel: "branch-redpitaya-v2017.2"
# Pro Applications: ""
##############################################################################
Last login: Fri Jul  7 01:20:40 2023 from 192.168.1.1
root@rp-f09e25:~# python3 time_osc_data.py
triggered
duration: 3.967 ms
triggered
duration: 3.609 ms
triggered
duration: 3.537 ms
triggered
duration: 3.530 ms
triggered
duration: 3.445 ms
triggered
duration: 3.502 ms
triggered
duration: 3.455 ms
triggered
duration: 3.529 ms
triggered
duration: 3.446 ms
triggered
duration: 3.539 ms
Overlay __del__ was activated.
But since there are garbage collection ordering issues, this code is commented out.
Exception ignored in: <bound method osc.__del__ of <redpitaya.overlay.mercury.mercury.osc object at 0xb69c8670>>
Traceback (most recent call last):
  File "/home/jupyter/RedPitaya/redpitaya/drv/osc.py", line 62, in __del__
  File "/home/jupyter/RedPitaya/redpitaya/drv/uio.py", line 90, in __del__
BufferError: cannot close exported pointers exist
```

**Comment by b-swat** [comment]
Hi.
I am not a python expert, the code that has changed to capture data is this one.
Most likely it is not the most optimal.

[osc changes](https://github.com/RedPitaya/jupyter/commit/52ffd7e92f3437381c238c90ad17a18a7c71a0d8#diff-b9bf3b3ea33f6e667bee333493169a86a2142b61ff83fe7f8a77f985b94f9dbaL137)

**Comment by spicydonkey** [comment]
Yes, that was exactly what was slowing things down.

Here is an implementation that is more efficient is the one currently in Beta 2.0.  This brings the speed for `data` down to 3 ms (so it is an upgrade over OS 1.04), with result below.

Here's the python code:
```python
    def data(self, siz: int = buffer_size, ptr: int = None) -> np.array:
        """Data.

        Parameters
        ----------
        siz : int, optional
            Number of data samples to be read from the FPGA buffer.
        ptr : int, optional
            End of data pointer, only use if you understand
            the source code.

        Returns
        -------
        array
            Array containing float samples scaled
            to the selected analog range.
            The data is alligned at the end to the last sample
            stored into the buffer.
        """
        if ptr is None:
            ptr = int(self.pointer)
        scale = self.__input_range / float(self._DWr)

        buffer_np = np.ctypeslib.as_array(self.buffer)
        buffer_size = len(self.buffer)

        # get `siz`-segment by start and exlusive-end pointers
        start = ptr
        end = (ptr + siz) % buffer_size
        if start < end:
            return scale * buffer_np[start:end]
        else:  # wrap around
            head_length = buffer_size - start
            segment = np.empty(head_length + end)
            segment[:head_length] = buffer_np[start:]
            segment[head_length:] = buffer_np[:end]
            return scale * segment
```

And here's the benchmark:
```
##############################################################################
# Red Pitaya GNU/Linux Ecosystem
# Version: 2.00-27dfa19da
# Build: 128
# Branch:
# Commit: 27dfa19da1cac87ae64dc6755bb4f729a6ffdfce
# U-Boot: "redpitaya-v2022.1"
# Linux Kernel: "branch-redpitaya-v2022.3"
# Pro Applications: ""
##############################################################################
Last login: Fri Jul  7 05:43:03 2023 from 192.168.1.1
root@rp-f09dc4:~# python3 time_osc_data.py
Check FPGA [OK].
triggered
duration: 3.567 ms
triggered
duration: 3.278 ms
triggered
duration: 3.125 ms
triggered
duration: 3.033 ms
triggered
duration: 3.013 ms
triggered
duration: 3.125 ms
triggered
duration: 3.028 ms
triggered
duration: 3.025 ms
triggered
duration: 3.041 ms
triggered
duration: 3.015 ms
Overlay __del__ was activated.
But since there are garbage collection ordering issues, this code is commented out.
Exception ignored in: <function osc.__del__ at 0xb006eb68>
Traceback (most recent call last):
  File "/home/jupyter/RedPitaya/redpitaya/drv/osc.py", line 61, in __del__
  File "/home/jupyter/RedPitaya/redpitaya/drv/uio.py", line 90, in __del__
BufferError: cannot close exported pointers exist
```

**Comment by spicydonkey** [comment]
Hey @b-swat , I'd like to submit the above change as a PR.  What do you think about that?

**Comment by b-swat** [comment]
Need to check and add:
https://github.com/RedPitaya/jupyter/tree/redpitaya_OS_2.x


According to the plans, there is no transfer of mercury to other boards. There are a lot of different board versions.
The plan is to make a python wrapper for rp-api and use the 0.94 version of FPGA. This code is unlikely to be useful in the future.

**Comment by b-swat** [comment]
@spicydonkey  

I have updated the code.
https://github.com/RedPitaya/jupyter/commit/6ad2849016f09203bebbbb028f4cdbf8b7895177


But I forgot one feature. In the new version of the ecosystem, I completely changed the calibration structure and the new calibration format is not applicable in Jupyter.
We still need to update the code for a new calibration, but there are no such plans.

**Comment by spicydonkey** [comment]
@b-swat Thanks for that.  I haven't come across the calibration structure in my (very short) experience with redpitaya.  How does the change in calibration structure affect the specification of the jupyter's `osc` module?

**Comment by b-swat** [comment]
There are two calibration versions in the latest update,
module that calculates the calibration
https://github.com/RedPitaya/jupyter/blob/redpitaya_OS_2.x/redpitaya/drv/clb.py
this module contains the old version and only for boards 125-14.
On other boards, the calibration structure is different.

If the calibration of another format is stored in the eeprom, then most likely the result of the work of osc and gen will not be correct.

**Comment by spicydonkey** [comment]
@b-swat thanks for the explanation.  Interesting thing is that I am using the nightly build you posted above, but have already been using the exactly same calibration file that you linked...
It could be because I "overwrite" all when unzipping the ecosystem.  But I haven't been finding any issues so far!

**Comment by b-swat** [comment]
the python source code for jupyter is not in the ecosystem. It is in the Linux image. Updating the ecosystem does not affect it in any way.

the code for jupyter is here (in large images):
https://downloads.redpitaya.com/downloads/Unify/RedPitaya_OS_2.00-15_beta.img.zip

And the changes that will appear in the distribution will appear only after I compile a new image.
